src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -170,6 +173,35 @@ internal Cci.IFieldReference GetModuleVersionId(Cci.ITypeReference mvidType)
             return _mvidField;
         }
 
+        internal Cci.IFieldReference GetInstrumentationPayload(int analysisKind, Cci.ITypeReference payloadType)
+        {
+            InstrumentationPayloadField payloadField;
+            if (!_instrumentationPayloadFields.TryGetValue(analysisKind, out payloadField))
+            {
+                payloadField = new InstrumentationPayloadField(this, analysisKind, payloadType);
```

Comment:
It feels like we need to do something to ensure that we are not trying to add fields after the type is frozen.